src/win/pipe.c

```diff
@@ -1962,29 +1965,50 @@ static DWORD uv__pipe_read_data(uv_loop_t* loop,
   if (max_bytes > buf.len)
     max_bytes = buf.len;
 
-  /* Read into the user buffer. */
-  if (!ReadFile(handle->handle, buf.base, max_bytes, &bytes_read, NULL)) {
-    uv__pipe_read_error_or_eof(loop, handle, GetLastError(), buf);
-    return 0; /* Break out of read loop. */
+  /* Read into the user buffer.
+   * Prepare an Event so that we can cancel if it doesn't complete immediately.
+   */
+  req = &handle->read_req;
+  memset(&req->u.io.overlapped, 0, sizeof(req->u.io.overlapped));
+  req->u.io.overlapped.hEvent = (HANDLE) ((uintptr_t) req->event_handle | 1);
+  if (ReadFile(handle->handle, buf.base, max_bytes, bytes_read, &req->u.io.overlapped)) {
+    r = ERROR_SUCCESS;
+  } else {
+    r = GetLastError();
+    *bytes_read = 0;
+    if (r == ERROR_IO_PENDING) {
+      r = CancelIoEx(handle->handle, &req->u.io.overlapped);
+      assert(r || GetLastError() == ERROR_NOT_FOUND);
```

Comment:
Wouldn't ERROR_NOT_FOUND after CancelIoEx indicate a bug?

Reply:
I believe ERROR_NOT_FOUND means that the operation finished (e.g. asynchronously) before libuv got a chance to call CancelIOEx, so that overlapped pointer was no longer found in the kernel's internal queues, but we still must wait for the kernel to release the associated resources (https://learn.microsoft.com/en-us/answers/questions/116109/cancelioex-spuriously-fails-with-error-not-found-w)

FWIW, this will almost never get called now, because (like unix) it breaks from the loop if it gets a short read. But I tested tweaking the loop logic to disable that break-on-short-read logic so that this did get stress tested and I observed this ERROR_IO_PENDING code path being exercised and working as expected.